x/ibc/07-tendermint/types/misbehaviour_handle_test.go

```diff
@@ -68,37 +68,65 @@ func (suite *TendermintTestSuite) TestCheckMisbehaviourAndUpdateState() {
 			true,
 		},
 		{
-			"invalid misbehavior misbehaviour from different chain",
+			"valid misbehavior misbehaviour with different trusted heights",
 			types.NewClientState(chainID, types.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, height, commitmenttypes.GetSDKSpecs(), false, false),
-			types.NewConsensusState(suite.now, commitmenttypes.NewMerkleRoot(tmhash.Sum([]byte("app_hash"))), height, bothValsHash),
-			types.NewConsensusState(suite.now, commitmenttypes.NewMerkleRoot(tmhash.Sum([]byte("app_hash"))), height, bothValsHash),
+			types.NewConsensusState(suite.now, commitmenttypes.NewMerkleRoot(tmhash.Sum([]byte("app_hash"))), heightMinus1, bothValsHash),
+			types.NewConsensusState(suite.now, commitmenttypes.NewMerkleRoot(tmhash.Sum([]byte("app_hash"))), heightMinus3, suite.valsHash),
 			&types.Misbehaviour{
-				Header1:  types.CreateTestHeader("ethermint", int64(height.EpochHeight), int64(height.EpochHeight), suite.now, bothValSet, bothValSet, bothSigners),
-				Header2:  types.CreateTestHeader("ethermint", int64(height.EpochHeight), int64(height.EpochHeight), suite.now.Add(time.Minute), bothValSet, bothValSet, bothSigners),
-				ChainId:  "ethermint",
+				Header1:  types.CreateTestHeader(chainID, height, heightMinus1, suite.now, bothValSet, bothValSet, bothSigners),
+				Header2:  types.CreateTestHeader(chainID, height, heightMinus3, suite.now.Add(time.Minute), bothValSet, suite.valSet, bothSigners),
+				ChainId:  chainID,
 				ClientId: chainID,
 			},
 			suite.now,
-			false,
+			true,
 		},
 		{
-			"valid misbehavior misbehaviour with different trusted heights",
-			types.NewClientState(chainID, types.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, height, commitmenttypes.GetSDKSpecs(), false, false),
+			"valid misbehaviour at a previous epoch",
+			types.NewClientState(chainIDEpoch1, types.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, clienttypes.NewHeight(1, 1), commitmenttypes.GetSDKSpecs(), false, false),
 			types.NewConsensusState(suite.now, commitmenttypes.NewMerkleRoot(tmhash.Sum([]byte("app_hash"))), heightMinus1, bothValsHash),
 			types.NewConsensusState(suite.now, commitmenttypes.NewMerkleRoot(tmhash.Sum([]byte("app_hash"))), heightMinus3, suite.valsHash),
 			&types.Misbehaviour{
-				Header1:  types.CreateTestHeader(chainID, epochHeight, epochHeight-1, suite.now, bothValSet, bothValSet, bothSigners),
-				Header2:  types.CreateTestHeader(chainID, epochHeight, epochHeight-3, suite.now.Add(time.Minute), bothValSet, suite.valSet, bothSigners),
-				ChainId:  chainID,
+				Header1:  types.CreateTestHeader(chainIDEpoch0, height, heightMinus1, suite.now, bothValSet, bothValSet, bothSigners),
+				Header2:  types.CreateTestHeader(chainIDEpoch0, height, heightMinus3, suite.now.Add(time.Minute), bothValSet, suite.valSet, bothSigners),
+				ChainId:  chainIDEpoch0,
+				ClientId: chainID,
+			},
+			suite.now,
+			true,
+		},
+		{
+			"valid misbehaviour at a future epoch",
```

Comment:
I find this very strange. This would mean a client could become frozen at a future epoch number:

`cs.FrozenHeight = tmMisbehaviour.GetHeight().(clienttypes.Height)`

I find this test confusing.

ClientState is at epoch 0. 
Misbehaviour has a epoch 0 in chainID
Misbehaviour header 1 has a height of epoch 1 and a trusted height of epoch 0
Misbehaviour header 2 has a height of epoch 1 and trusted height of epoch 0.

The test expects to pass, and the client at height 0-4 now has a frozen height of 1-3? This is incorrect behaviour right?   I don't think we can punish for future header misbehaviour since we cannot know if the light client would have been fooled given its params?

Shouldn't we enforce that the misbehaviour chainID == the misbehaviour height if in the epoch format?

Reply:
cc @cwgoes maybe you could clarify how this should work?

Reply:
bump @cwgoes

Reply:
Ah sorry. Thanks for bringing this up. We do want to cover the case where the client hasn't yet been updated to the new epoch but if it were updated it would have been fooled (within the unbonding window). I guess that should require an additional proof of the `upgrade` function call itself (the new client state written at the appropriate key). in which case `upgrade` could be called before misbehaviour was submitted, and we could safely enforce that misbehaviour chain ID & height are equal. Does that reasoning seem right?

Reply:
Concluded that it is safe to accept future epoch misbehavior, although it is possible that the light client would not have been fooled if `upgrade` was never called.